design/tlog-spilling.md.html

```diff
@@ -0,0 +1,590 @@
+performance.  A write bandwidth saturation test was run against a cluster, with
+a modification to the transaction logs to have them act as if there was one
+storage server that was permanently failed; it never sent pop requests to allow
+the TLog to remove data from memory.  After 15min, the write bandwidth had
+reduced to 30% of its baseline.  After 30min, that became 10%.  After 60min,
+that became 5%.  (This is an intentional hyperbole due to the saturating write
+load.  See experiments at the end for more detail and data.)
+
+With the recent multi-DC/multi-region work, a failure of a remote data center
+would cause transaction logs to need to buffer all commits, as every commit is
+tagged as destined for the remote datacenter.  This would rapidly push
+transaction logs into a spilling regime, and thus write bandwidth would begin
+to rapidly degrade.  It is unacceptable for a remote datacenter failure to so
+drastically affect the primary datacenter's performance in the case of a
+failure, so a more performant way of spilling data is required.
+
+## Overview
+
+Whereas spill-by-value copied the entire mutation into the B-tree and removes
+it from the disk queue, spill-by-reference leaves the mutations in the disk
+queue and writes a pointer to it into the B-tree. Performance experiments
+revealed that the TLog's performance while spilling was dictated more by the
+number of writes done to the SQLite B-tree, than by the size of those writes.
+Thus, "spill-by-reference" being able to do a significantly better batching
+with its writes to the B-tree is more important than that it writes less data
+in aggregate.  Spill-by-reference significantly reduces the volume of data
+written to the B-tree, and the less data that we write, the more we can batch
+versions to be written together.
+
+************************************************************************
+*                              DiskQueue                               *
+*                                                                      *
+*    -------- Index on disk --------     ---- Index in memory ----     *
+*   /                               \   /                         \    *
+* +-----------------------------------+-----------------------------+  *
+* |            Spilled Data           |       Most Recent Data      |  *
+* +-----------------------------------+-----------------------------+  *
+* lowest version                                      highest version  *
+*                                                                      *
+************************************************************************
+
+Spill-by-reference works by taking a larger range of versions, and building a
+single key-value pair per tag that describes where in the disk queue is every
+relevant commit for that tag.  Concretely, this takes the form
+`(tag, last_version) -> [(version, start, end, mutation bytes)]`, where...
```

Comment:
Are there multiple entries per `(tag, last_version)` I.E does the value for one key contain an array of tuples in the form `(version, start, end, mutation bytes)` ? If so it might be more clear to represent the value like this: `[(version, start, end, mutation bytes), ...]` where the `...` signals there are more than one entry

Reply:
Done